### ISSUE TITLE
Improve performance by checking uuid first

### ISSUE THREAD

**Issue by rwestlund** [iocage/iocage#244]
As discussed by @skarekrow and @pmhausen in #230, iocage creates a list of jails before running any command in order to do tag matching.

Here's an idea for speeding up single-jail commands during jail matching: do a first pass only matching the full uuid/name against the list of dataset names.  If that doesn't find anything, only then read in all the jail config files in order to match against tags.

I haven't measured where iocage is spending its time, but I think this would be a significant performance boost for large deployments.

**Comment by skarekrow** [comment]
That's actually how we do it currently. The configuration is loaded later. But I'll mark this as an enhancement and change the title if my idea pans out

**Comment by pmhausen** [comment]
Hey, guys,

> As discussed by @skarekrow and @pmhausen in #230, iocage creates a list of jails before running any command in order to do tag matching.
> 
> Here's an idea for speeding up single-jail commands during jail matching: do a first pass only matching the full uuid/name against the list of dataset names. If that doesn't find anything, only then read in all the jail config files in order to match against tags.
> 
> I haven't measured where iocage is spending its time,

But I did ;-)

Thousands of ioctl() against /dev/zfs - i.e. probably reading ZFS attributes.
I have not looked at your code, Brandon, but wouldn't it be possible to just
read the config files for most operations?

I.e. why do you need a list of anything from ZFS for `iocage console foo`?

-	read /where/iocage/lies/jails/*/config.json to find the one with UUID or
	tag "foo"
-	jexec $foo-jail /bin/csh

Possibly I'm not seeing the elephant here, but what do you need from
ZFS for console, exec, start, stop ...?

Patrick

**Comment by rwestlund** [comment]
@skarekrow the configuration is not loaded later.  See this example:

```
root@r2d2 /u/h/randy# iocage create -n test1 -r 11.0-RELEASE
test1 (test1) successfully created!

root@r2d2 /u/h/randy# iocage create -n test2 -r 11.0-RELEASE
test2 (test2) successfully created!

root@r2d2 /u/h/randy# rm /iocage/jails/test2/config.json

root@r2d2 /u/h/randy# iocage start test1
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/iocage/lib/ioc_json.py", line 187, in json_load
    with open(self.location + "/config.json", "r") as conf:
FileNotFoundError: [Errno 2] No such file or directory: '/iocage/jails/test2/config.json'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/iocage", line 10, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/iocage/cli/start.py", line 55, in cli
    ioc.IOCage(jail, rc=rc).start()
  File "/usr/local/lib/python3.6/site-packages/iocage/lib/iocage.py", line 100, in __init__
    self.jails, self._paths = self.list("uuid")
  File "/usr/local/lib/python3.6/site-packages/iocage/lib/iocage.py", line 896, in list
    plugin=plugin, quick=quick).list_datasets()
  File "/usr/local/lib/python3.6/site-packages/iocage/lib/ioc_list.py", line 82, in list_datasets
    conf = iocage.lib.ioc_json.IOCJson(jail).json_load()
  File "/usr/local/lib/python3.6/site-packages/iocage/lib/ioc_json.py", line 276, in json_load
    self.json_convert_from_zfs(uuid, skip=skip)
UnboundLocalError: local variable 'uuid' referenced before assignment
```

iocage is reading the configuration for all jails before it tries to match the uuid/name or tag against any of them, which is unnecessary.  In this example, `iocage start test1` could have seen that `test1` is an exact uuid/name match and not bothered to load the config for `test2` and the other jails; it shouldn't have noticed that `test2`'s `config.json` was missing.

The code doing it is here:
https://github.com/iocage/iocage/blob/082a591d6910700b82973d8e82912d307e240749/iocage/lib/ioc_list.py#L80-L82

**Comment by skarekrow** [comment]
@rwestlund hmm it shouldn't be doing that. But I'll be honest and say it was one of the first things I did in both python and this rewrite. So it's a nasty hairball I've mostly blocked out (it was traumatic) and @pmhausen that's precisely what I was thinking about doing. But the issue is that the datasets don't match a tags name only the uuid. 

I was toying with the idea of renaming the config.json to the jails tag.json and the dataset to be uuid to make it a quick filter through the directories. But that has its own issues.

I think what we need is a dual stage match, IE the fastest one like @rwestlund suggests and a deeper search if nothing was found.

I grab the datasets so I can dynamically get their path @pmhausen and match them to the UUID. As you can change the mountpoint at anytime but I think that can be done much smarter. 

So TLDR ; this function sucks and certainly can be improved. I'll see to it :)

**Comment by rwestlund** [comment]
The real problem IMO is the old uuid-and-tag architecture -- in the case of a tag, it forces you to scan every jail's mountpoint and config before you know which jail the user wants to operate on.  iocage could be several orders of magnitude faster (and _O(1)_ rather than _O(n)_) if it didn't allow specifying jails by tag or partial uuid/name matching.  You wouldn't even have to iterate over all the datasets; you'd just go directly to the provided uuid/name.

Without making that architecture change, there's only so much @skarekrow can do.  iocage's performance is always going to suck when you have a significant number of jails :(

You could make such an architecture change easier for users by having a command for "which jails have a tag/description with this value?". `iocage search tag=mytag` and `iocage search ip4_addr="em0|192.168.1.234/32"`

**Comment by skarekrow** [comment]
Yeah that's really the skinny of it. Without a major shift (which changes iocage fundamentally to it's former years) I'm a bit hamstrung. That being said, the `search` command is a good idea, it's essentially like `get -r` but gives you back a jail. This could possibly just be a flag to `get` like `-s` or something if I haven't used that yet.

That being said, since the introduction of the `--name` function, I think the usefulness of tags is possibly NIL. It gives me a ton of headaches (like #238) and isn't as intuitive as it seems. By that I mean one would assume you are actually *tagging* the jail, and not needing it to be unique, but rather a metadata like `database; postgres` that many jails can share for you to filter and drill down. I don't think the tags being unique and a replacement is as elegant solution as I'd like.

So having thought about this, and I did ask in that other ticket, I think this is what I'm actually going to do with iocage:
```
1 - Remove TAGs (redundant if you want that functionality use --name, perhaps I flip this default though?)
2 - Rip out the duplicate tag detection
3 - Rip out the `/iocage/tags` infrastructure entirely
4 - Change tags to some metadata tagging instead of essentially a symlink
5 - Profit??
```

Thoughts? I think this could be a major feature for 0.9.9 and a lovely thing to introduce before 1.0 where this breakage will not be acceptable until a 1.1 or later.


EDIT: This also brings the question of migration, and if so which one do we keep. UUID or TAG? I'm leaning towards migration using the TAG as the name if it doesn't equal a date, and the UUID if it does. Since that means you didn't supply a TAG and you used the UUID instead.

This allows us to then adjust the `host_hostuuid` property to match that, rename the datasets and we're done. Not too bad of a process, and the user may never notice since they always used the TAG/UUID and all the informational messages talking about the UUID and TAG can be dramatically shortened.

Overall this seems like a major win that we should chase, but before I shit in the sandbox, I want to make sure I'm not too off the needs of you guys and this would be a net benefit.

**Comment by rwestlund** [comment]
That's a really good point about tags.  I agree that they're no longer useful, and I think removing them is a good call.

Also note that once you remove tags, there will no longer be a difference between the `-u` and `-n` options to `create` and `clone`, so that can be simplified as well.

I realized that you can still support partial name matching without much of a performance hit by first checking to see whether a dataset with the exact name exists, and only iterating over all the datasets for partial matches if the exact name doesn't exist.

**Comment by skarekrow** [comment]
Yeah really we just miss out on two systems for naming.  It as you note allows a much faster access to the jails and simplifies everything. I think the positives outweigh the negatives. It could be night and day on systems. 

I really think I'm going to forge ahead with this. Could be something really beneficial to `iocage`

**Comment by igalic** [comment]
this is an interesting development!

i was gonna ask how to prevent `iocage create` from a cage with a $DATE tag when the tag in question already exists.

**Comment by skarekrow** [comment]
Good news, made significant progress with this. Will continue and should hopefully have this done before the end of the week 👍 

EDIT: `#4` of my list is not going to be included with this, that's a possible feature in the long run, but not on the table for now.

**Comment by igalic** [comment]
changing tags to be… well, tags… and, multiple thereof, too, would make them much more useful, imo

**Comment by skarekrow** [comment]
Yeah that's far more ambitious and time consuming. It's already a big hairball to unravel let alone making them actually tags like I suggested. That's absolutely a future project though

**Comment by skarekrow** [comment]
@pmhausen Ok all pushed, my tests show everything to be much faster! Lemme know 👍

**Comment by rwestlund** [comment]
Check it out!  Here's a before and after on a system with 1000 dummy jails:

Before:
```
root@r2d2 /u/h/randy# time iocage start test
* Starting test (test)
  + Started OK
  + Starting services OK
      109.67 real         4.61 user        57.93 sys
```
After:
```
root@r2d2 /u/h/randy# time iocage start test
* Starting test
  + Started OK
  + Starting services OK
        4.63 real         0.94 user         3.08 sys
```

**Comment by pmhausen** [comment]
Hi, Brandon,

could you take the time to once more look into
https://bugs.freenas.org/issues/25169, please?

Thanks!
Patrick
